geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/services/StripedExecutorService.java

```diff
@@ -0,0 +1,522 @@
+        ser_ex.tasks.drainTo(result);
+      }
+      result.addAll(executor.shutdownNow());
+      return result;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Returns true if shutdown() or shutdownNow() have been
+   * called; false otherwise.
+   */
+  public boolean isShutdown() {
+    lock.lock();
+    try {
+      return state == State.SHUTDOWN;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Returns true if this pool has been terminated, that is, all
+   * the SerialExecutors are empty and the wrapped
+   * ExecutorService has been terminated.
+   */
+  public boolean isTerminated() {
+    lock.lock();
+    try {
+      if (state == State.RUNNING)
+        return false;
+      for (SerialExecutor executor : executors.values()) {
+        if (!executor.isEmpty())
+          return false;
+      }
+      return executor.isTerminated();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Returns true if the wrapped ExecutorService terminates
+   * within the allotted amount of time.
+   */
+  public boolean awaitTermination(long timeout, TimeUnit unit)
+      throws InterruptedException {
+    lock.lock();
+    try {
+      long waitUntil = System.nanoTime() + unit.toNanos(timeout);
+      long remainingTime;
+      while ((remainingTime = waitUntil - System.nanoTime()) > 0
+          && !executors.isEmpty()) {
+        terminating.awaitNanos(remainingTime);
+      }
+      if (remainingTime <= 0)
+        return false;
+      if (executors.isEmpty()) {
+        return executor.awaitTermination(
+            remainingTime, TimeUnit.NANOSECONDS);
+      }
+      return false;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * As soon as a SerialExecutor is empty, we remove it from the
+   * executors map. We might thus remove the SerialExecutors
+   * more quickly than necessary, but at least we can avoid a
+   * memory leak.
+   */
+  private void removeEmptySerialExecutor(Object stripe, SerialExecutor ser_ex) {
+    assert ser_ex == executors.get(stripe);
+    assert lock.isHeldByCurrentThread();
+    assert ser_ex.isEmpty();
+
+    executors.remove(stripe);
+    terminating.signalAll();
+    if (state == State.SHUTDOWN && executors.isEmpty()) {
+      executor.shutdown();
+    }
+  }
+
+  /**
+   * Prints information about current state of this executor, the
+   * wrapped executor and the serial executors.
+   */
+  public String toString() {
+    lock.lock();
+    try {
+      return "StripedExecutorService: state=" + state + ", " +
+          "executor=" + executor + ", " +
+          "serialExecutors=" + executors;
+    } finally {
+      lock.unlock();
+    }
+
+  }
+
+  /**
+   * This field is used for conditional compilation. If it is
+   * false, then the finalize method is an empty method, in
+   * which case the SerialExecutor will not be registered with
+   * the Finalizer.
+   */
+  private static boolean DEBUG = false;
```

Comment:
This can be `final`.

Reply:
I've removed all this debug stuff since we wouldn't have any means of switching it on if we thought there was a problem. We'd have to introduce a system variable to set it externally.